File: apps/web/lib/load.ts
// The typed Load shared across the plan and the record (ADR-0010). A Load is no
// longer a bare string: its `kind` fixes how it resolves to a number, and `text`
// carries the original, display-ready free text. This module has NO server-only
// imports, so it is safe in both Server and Client Components.

import type { WeightUnit } from "./weight-unit";
import {
  formatWeight,
  formatWeightNumber,
  unitToKg,
  weightUnitLabel,
} from "./weight-format.ts";

export type LoadKind =
  | "absolute"
  | "bodyweight"
  | "percent_1rm"
  | "qualitative"
  | "range";

// The wire shape the API serializes for a typed Load. Only the payload fields the
// `kind` carries are present; `text` is always there and is what the UI displays.
export interface Load {
  kind: LoadKind;
  text: string;
  kg?: number;
  added_kg?: number;
  percent?: number;
  low_kg?: number;
  high_kg?: number;
}

// The em dash the UI shows when no Load was prescribed or logged.
export const NO_LOAD = "—";

// Whether a Load is **pure bodyweight** — the one axis Progression steps by reps (ADR-0026),
// and so the only movement a rep Pin governs (ADR-0053). The web mirror of the backend
// `_is_pure_bodyweight`: the BODYWEIGHT kind carrying no added load. A weighted-bodyweight
// (added-load), loaded, %-1RM, or load-less movement progresses on Load, not reps. `added_kg`
// is omitted on the wire for pure bodyweight, so a nullish check is the "no added load" test.
export function isPureBodyweight(load: Load | null | undefined): boolean {
  return load?.kind === "bodyweight" && load.added_kg == null;
}

// Render a typed Load for display in the reader's Weight Unit, falling back to the em dash
// when absent. The weight-bearing kinds are computed from their numeric **kilogram** fields —
// `absolute` from `kg`, weighted `bodyweight` from `added_kg`, `range` from its bounds — and
// projected to the reader's unit through the `weight-format` seam, so a Redeemed / Shared
// Session renders in the recipient's unit (#417). Pure bodyweight (no added load), `percent_1rm`,
// and `qualitative` carry no kg number and are unit-agnostic, so their preserved `text` stays
// authoritative. A weight-bearing kind missing its number falls back to `text` rather than
// rendering a broken string.
export function formatLoad(
  load: Load | null | undefined,
  unit: WeightUnit,
): string {
  if (load == null) return NO_LOAD;
  switch (load.kind) {
    case "absolute":
      return load.kg != null ? formatWeight(load.kg, unit) : load.text ?? NO_LOAD;
    case "bodyweight":
      return load.added_kg != null
        ? `bodyweight + ${formatWeight(load.added_kg, unit)}`
        : load.text ?? NO_LOAD;
    case "range":
      return load.low_kg != null && load.high_kg != null
        ? `${formatWeightNumber(load.low_kg, unit)}-${formatWeightNumber(load.high_kg, unit)} ${weightUnitLabel(unit)}`
        : load.text ?? NO_LOAD;
    default:
      return load.text ?? NO_LOAD;
  }
}

// Render a Performed Body Weight (ADR-0026) for display in the reader's Weight Unit — the
// captured mass projected from canonical kilograms, or the em dash when none was on file
// (never guessed). Shares the `weight-format` seam so the body-weight stat tracks the same
// unit projection as every other weight (#417).
export function formatBodyWeight(
  kg: number | null | undefined,
  unit: WeightUnit,
): string {
  return kg != null ? formatWeight(kg, unit) : NO_LOAD;
}

// Reverse a typed Load into a picker's `loadKind` + `loadValue` (ADR-0010) — the raw field
// the picker sends for the kind, expressed in the reader's Weight Unit so the input shows what
// they would type: the weight for absolute, the percent for `percent_1rm`, the added weight for
// bodyweight (blank for pure bodyweight), the `low-high` pair for a range, the free text for
// qualitative. The kilogram fields are projected to `unit` through the `weight-format` seam
// (rounded at the display boundary, so a round-trip through the input is drift-free, #417);
// percent and qualitative are unit-agnostic. No Load → a blank absolute field. The single
// source of truth shared by the Log Correction pre-fill and the Capture seed, so the two
// reverse mappings never drift.
export function loadToFields(
  load: Load | null,
  unit: WeightUnit,
): { loadKind: LoadKind; loadValue: string } {
  if (load === null) return { loadKind: "absolute", loadValue: "" };
  switch (load.kind) {
    case "absolute":
      return {
        loadKind: "absolute",
        loadValue: load.kg != null ? formatWeightNumber(load.kg, unit) : "",
      };
    case "percent_1rm":
      return {
        loadKind: "percent_1rm",
        loadValue: load.percent != null ? String(load.percent) : "",
      };
    case "bodyweight":
      return {
        loadKind: "bodyweight",
        loadValue: load.added_kg != null ? formatWeightNumber(load.added_kg, unit) : "",
      };
    case "range":
      return {
        loadKind: "range",
        loadValue:
          load.low_kg != null && load.high_kg != null
            ? `${formatWeightNumber(load.low_kg, unit)}-${formatWeightNumber(load.high_kg, unit)}`
            : "",
      };
    default:
      return { loadKind: "qualitative", loadValue: load.text ?? "" };
  }
}

// The pattern a range value carries — `low-high`, whitespace-tolerant, decimal bounds — so
// each bound can be converted independently. Mirrors the backend range grammar.
const RANGE_VALUE = /^\s*(\d*\.?\d+)\s*-\s*(\d*\.?\d+)\s*$/;

// Convert a picker's `loadValue` — typed in the reader's Weight Unit — into the canonical
// **kilogram** string the backend's `load_from_input` parses (which reads a Load value as kg).
// This is the input half of the unit projection (#417): the weight-bearing kinds (`absolute`,
// `bodyweight` added load, `range` bounds) are converted to exact kilograms with no rounding,
// so what is stored round-trips to the entered value; `percent_1rm` and `qualitative` are
// unit-agnostic and pass through untouched. When the reader's unit is already kilograms the
// value passes through verbatim, so a kg user's submission is byte-for-byte unchanged. A blank
// or unparseable value passes through so the existing backend fallbacks still handle it.
export function loadValueToKg(
  loadKind: string,
  loadValue: string,
  unit: WeightUnit,
): string {
  if (unit === "kg") return loadValue;
  const raw = loadValue.trim();
  if (raw === "") return loadValue;
  switch (loadKind) {
    case "absolute":
    case "bodyweight": {
      const value = Number(raw);
      return Number.isFinite(value) ? String(unitToKg(value, unit)) : loadValue;
    }
    case "range": {
      const pair = RANGE_VALUE.exec(raw);
      if (pair === null) return loadValue;
      return `${unitToKg(Number(pair[1]), unit)}-${unitToKg(Number(pair[2]), unit)}`;
    }
    default:
      return loadValue;
  }
}

// The kinds offered by the log form's picker, paired with a human label. The absolute kind's
// label carries the reader's active Weight Unit — "Weight (kg)" / "Weight (lb)" — replacing the
// old hardcoded "(kg)" so the picker names the unit the field is actually entered in (#417).
export function loadKindOptions(
  unit: WeightUnit,
): ReadonlyArray<{ value: LoadKind; label: string }> {
  return [
    { value: "absolute", label: `Weight (${weightUnitLabel(unit)})` },
    { value: "bodyweight", label: "Bodyweight" },
    { value: "percent_1rm", label: "% of 1RM" },
    { value: "range", label: "Range" },
    { value: "qualitative", label: "Descriptive" },
  ];
}

File: apps/api/app/domain/load.py
"""Load — a typed value object for the prescribed/performed weight (ADR-0010).

``parse_load`` is the single source of truth for interpreting the free-text load
vocabulary. It is a pure function: given the raw string the AI prescribed or the
user logged, it returns a tagged ``ParsedLoad`` whose ``kind`` fixes the *meaning*
of the load once, at the write boundary, so downstream analytics never re-guess.

The free-text-ness is essential domain vocabulary, not an accident: the AI
legitimately prescribes ``bodyweight`` and ``70% 1RM``, so those are valid kinds,
not malformed data. Only some kinds resolve to a number — ``absolute`` carries a
kilogram value directly; ``bodyweight`` and ``percent_1rm`` resolve only against
the user's mass or estimated 1RM (elsewhere); ``range`` carries a low/high; and
``qualitative`` never resolves. Every ``ParsedLoad`` also keeps the original
``text`` verbatim, so the typed representation is display-ready and a backfill
loses nothing.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import Enum


class LoadKind(str, Enum):
    """How a Load is expressed — and therefore how (or whether) it resolves to kg."""

    ABSOLUTE = "absolute"
    BODYWEIGHT = "bodyweight"
    PERCENT_1RM = "percent_1rm"
    QUALITATIVE = "qualitative"
    RANGE = "range"


_NUMBER = r"\d+(?:\.\d+)?"


def _to_number(raw: str) -> float:
    value = float(raw)
    return value


def _format_number(value: float) -> str:
    """Render a load number without a trailing ``.0`` (``70.0`` → ``"70"``)."""

    return str(int(value)) if value == int(value) else str(value)


@dataclass(frozen=True)
class ParsedLoad:
    """A load whose ``kind`` fixes its meaning, with the payload that kind carries.

    ``text`` is the original free-text, preserved verbatim so the value is both
    display-ready and lossless. The numeric payload fields are populated only for
    the kinds that carry them: ``kg`` for ``absolute``; ``added_kg`` for the
    optional extra load on a ``bodyweight`` movement; ``percent`` for
    ``percent_1rm``; ``low_kg``/``high_kg`` for a ``range``. ``qualitative`` carries
    no number.
    """

    kind: LoadKind
    text: str
    kg: float | None = None
    added_kg: float | None = None
    percent: float | None = None
    low_kg: float | None = None
    high_kg: float | None = None

    def to_dict(self) -> dict:
        """The JSON-column representation: ``kind`` and ``text`` always, plus only
        the numeric payload fields the kind actually carries."""

        data: dict = {"kind": self.kind.value, "text": self.text}
        for field_name in ("kg", "added_kg", "percent", "low_kg", "high_kg"):
            value = getattr(self, field_name)
            if value is not None:
                data[field_name] = value
        return data

    @classmethod
    def from_dict(cls, data: dict) -> "ParsedLoad":
        """Rebuild a ``ParsedLoad`` from its stored ``to_dict`` representation."""

        return cls(
            kind=LoadKind(data["kind"]),
            text=data["text"],
            kg=data.get("kg"),
            added_kg=data.get("added_kg"),
            percent=data.get("percent"),
            low_kg=data.get("low_kg"),
            high_kg=data.get("high_kg"),
        )


def parse_load(raw: str) -> ParsedLoad:
    """Interpret a raw free-text load into a tagged ``ParsedLoad``.

    A bare number (``"60"``) is an absolute kilogram load.
    """

    text = raw.strip()

    if re.search(r"body\s*weight|(?<![a-z])bw(?![a-z])", text, re.IGNORECASE):
        added = re.search(rf"({_NUMBER})\s*(?:kg|kgs|k)?", text, re.IGNORECASE)
        return ParsedLoad(
            kind=LoadKind.BODYWEIGHT,
            text=raw,
            added_kg=_to_number(added.group(1)) if added is not None else None,
        )

    percent = re.match(rf"({_NUMBER})\s*%", text)
    if percent is not None:
        return ParsedLoad(
            kind=LoadKind.PERCENT_1RM, text=raw, percent=_to_number(percent.group(1))
        )

    range_match = re.match(rf"({_NUMBER})\s*-\s*({_NUMBER})", text)
    if range_match is not None:
        return ParsedLoad(
            kind=LoadKind.RANGE,
            text=raw,
            low_kg=_to_number(range_match.group(1)),
            high_kg=_to_number(range_match.group(2)),
        )

    absolute = re.fullmatch(rf"({_NUMBER})\s*(?:kg|kgs|k)?", text, re.IGNORECASE)
    if absolute is not None:
        return ParsedLoad(kind=LoadKind.ABSOLUTE, text=raw, kg=_to_number(absolute.group(1)))

    return ParsedLoad(kind=LoadKind.QUALITATIVE, text=raw)


def resolve_bodyweight_kg(
    parsed: ParsedLoad, performed_body_weight: float | None
) -> float | None:
    """Resolve a ``bodyweight`` Load to its kg-equivalent against a captured mass.

    The single source of truth for bodyweight resolution (ADR-0026): a pure
    ``bodyweight`` set resolves to the Performed Body Weight, and a ``bodyweight +
    added`` set to mass plus the added load. Returns ``None`` when the Load is not a
    bodyweight kind, or when no Performed Body Weight was captured — the mass is never
    guessed, so a set with no snapshot stays unresolved (and therefore unscored)
    rather than being resolved against a fabricated weight.
    """

    if parsed.kind is not LoadKind.BODYWEIGHT:
        return None
    if performed_body_weight is None:
        return None
    return performed_body_weight + (parsed.added_kg or 0.0)


def load_from_input(kind: str, value: str | None) -> ParsedLoad | None:
    """Build a typed Load from the log form's kind picker plus its value field.

    The picked ``kind`` is authoritative — the user chose it deliberately, so a
    value of ``"70"`` under a ``percent_1rm`` pick is a percentage, not the
    absolute kilogram figure a bare parse would guess. A blank value is *no load*
    (``None``) for every kind except ``bodyweight``, which is itself a complete
    load even with no added weight. A value that cannot be read as the kind's
    number falls back to :func:`parse_load` rather than raising.
    """

    load_kind = LoadKind(kind)
    raw = (value or "").strip()

    if load_kind is LoadKind.BODYWEIGHT:
        if not raw:
            return ParsedLoad(kind=LoadKind.BODYWEIGHT, text="bodyweight")
        try:
            added = _to_number(raw)
        except ValueError:
            return parse_load(raw)
        return ParsedLoad(
            kind=LoadKind.BODYWEIGHT,
            text=f"bodyweight + {_format_number(added)} kg",
            added_kg=added,
        )

    if not raw:
        return None

    if load_kind is LoadKind.QUALITATIVE:
        return ParsedLoad(kind=LoadKind.QUALITATIVE, text=raw)

    if load_kind is LoadKind.RANGE:
        pair = re.match(rf"({_NUMBER})\s*-\s*({_NUMBER})", raw)
        if pair is None:
            return parse_load(raw)
        low, high = _to_number(pair.group(1)), _to_number(pair.group(2))
        return ParsedLoad(
            kind=LoadKind.RANGE,
            text=f"{_format_number(low)}-{_format_number(high)} kg",
            low_kg=low,
            high_kg=high,
        )

    try:
        number = _to_number(raw)
    except ValueError:
        return parse_load(raw)

    if load_kind is LoadKind.PERCENT_1RM:
        return ParsedLoad(
            kind=LoadKind.PERCENT_1RM, text=f"{_format_number(number)}% 1RM", percent=number
        )
    return ParsedLoad(
        kind=LoadKind.ABSOLUTE, text=f"{_format_number(number)} kg", kg=number
    )


__all__ = [
    "LoadKind",
    "ParsedLoad",
    "parse_load",
    "load_from_input",
    "resolve_bodyweight_kg",
]
